extensions/lite/queue/index.js

```diff
@@ -20,30 +38,48 @@ export default class Queue {
     } else {
       this.queue.push(jobWrapped)
     }
-    this.ee.emit('newJob')
+    this.tick()
   }
 
   dequeue() {
     return this.queue.shift()
   }
 
-  subscribe(fn) {
-    this.ee.on('newJob', async () => {
-      const { job, retries } = this.dequeue()
-      try {
-        await fn(job)
-      } catch (err) {
-        this.logger.error(`${this.name} queue failed to process job: ${err.message}`)
-
-        if (retries + 1 <= this.options.retries) {
-          this.enqueue(job, retries + 1, true)
-        } else {
-          this.logger.warn(
-            `[WARNING] Retrying job within ${this.name} queue failed
-            ${this.options.retries} times! Dropping it!`
-          )
-        }
-      }
+  async tick() {
+    const toDequeueIdx = _.findIndex(this.queue, el => {
+      const queueId = this.getQueueId(el.job)
+      return this._lock[queueId] !== true
     })
+
+    if (toDequeueIdx === -1) {
+      return
+    }
+
+    const [{ job, retries }] = _.pullAt(this.queue, [toDequeueIdx])
+    const queueId = this.getQueueId(job)
+    this._lock[queueId] = true
+
+    try {
+      await Promise.mapSeries(this.subscribers, fn => fn(job))
+    } catch (err) {
+      this.logger.warn(`${this.name} queue failed to process job: ${err.message}`)
+
+      if (retries + 1 <= this.options.retries) {
+        this.enqueue(job, retries + 1, true)
+      } else {
+        this.logger.error(
+          `Retrying job within ${this.name} queue failed ${this.options.retries} times. Abandoning the job.`
+        )
+      }
+    } finally {
```

Comment:
Wow, I think I have never ever seen `finally` in JS code o_O

Reply:
Haha right. Only useful in locking and garbage collection situations which are both extremely rare in JS